From 200fbf0c59369304a6e7cfe188b7d98e96b90af6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Florian=20M=C3=BCllner?= <fmuellner@gnome.org>
Date: Fri, 16 Jun 2017 17:08:07 +0200
Subject: [PATCH] system: Emulate click action button release

Since commit 2c070d38, we add a ClickAction to the visible AltSwitcher
button to track long-presses. As a result, we now have two components
that will grab and ungrab the pointer for the button, so to make sure
we don't end up with a stuck grab, we need to release the second's
component grab when the first activates.

Currently we only drop the StButton grab on long-press, we also need
to cancel any initiated long-press on click.

https://bugzilla.gnome.org/show_bug.cgi?id=781738
---
 js/ui/status/system.js | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/js/ui/status/system.js b/js/ui/status/system.js
index 02b683075..c2d197ace 100644
--- a/js/ui/status/system.js
+++ b/js/ui/status/system.js
@@ -42,9 +42,15 @@ const AltSwitcher = new Lang.Class({
     _init: function(standard, alternate) {
         this._standard = standard;
         this._standard.connect('notify::visible', Lang.bind(this, this._sync));
+        if (this._standard instanceof St.Button)
+            this._standard.connect('clicked',
+                                   () => { this._clickAction.release(); });
 
         this._alternate = alternate;
         this._alternate.connect('notify::visible', Lang.bind(this, this._sync));
+        if (this._alternate instanceof St.Button)
+            this._alternate.connect('clicked',
+                                    () => { this._clickAction.release(); });
 
         this._capturedEventId = global.stage.connect('captured-event', Lang.bind(this, this._onCapturedEvent));
 
-- 
2.13.0